### PR TITLE
Add SyncThreads primitive

### PR DESCRIPTION
**Pull request by haozech**
此PR添加了新的SyncThreads原语实现和测试

### CODE REVIEW

File: cinn/ir/ir_schedule.cc [hunk 1]
```diff
@@ -1201,6 +1201,58 @@ Expr IRSchedule::CacheWrite(const Expr& block, int write_buffer_index, const std
   return *find_cache_block.begin();
 }
 
+struct InsertExpr : public ir::IRMutator<> {
+ public:
+  static void Insert(Expr* expr, const Expr& ir_node, const Expr& insert_node, bool after_node) {
```
Comment (zhhsplendid) [inline cinn/ir/ir_schedule.cc]: Usually, input parameters are before output parameters.

Reference: https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

File: cinn/ir/ir_schedule.h [hunk 2]
```diff
@@ -216,6 +216,13 @@ class IRSchedule {
    */
   Expr CacheWrite(const Expr& block, int write_buffer_index, const std::string& memory_type);
 
+  /**
+   * \brief Add SyncThreads statements in AST.
+   * @param ir_node The insertion point in AST.
+   * @param after_node Whether to insert the statement after the insertion point.
```
Comment (zhhsplendid) [inline cinn/ir/ir_schedule.h]: So, we will insert the statement before the insertion point when it is False? Suggest to explain the two cases.

File: cinn/ir/ir_schedule.cc [hunk 3]
```diff
@@ -1201,6 +1201,58 @@ Expr IRSchedule::CacheWrite(const Expr& block, int write_buffer_index, const std
   return *find_cache_block.begin();
 }
 
+struct InsertExpr : public ir::IRMutator<> {
```
Comment (zhhsplendid) [inline cinn/ir/ir_schedule.cc:1204]: Is it possible that InsertExpr can be a tool used somewhere else? If yes, suggest to put it at some IR utility file. Our ir_schedule.cc is getting long now.

File: cinn/ir/ir_schedule.cc [hunk 4]
```diff
@@ -1201,6 +1201,58 @@ Expr IRSchedule::CacheWrite(const Expr& block, int write_buffer_index, const std
...
+
+ private:
+  explicit InsertExpr(const Expr& ir_node, const Expr& insert_node, bool after_node)
+      : ir_node_(ir_node), insert_node_(insert_node), after_node_(after_node) {}
+
+  void Visit(const ir::Block* expr, Expr* op) override {
+    for (int i = 0; i < expr->stmts.size(); i++) {
+      if (expr->stmts[i] == ir_node_) {
+        if (after_node_)
+          op->As<ir::Block>()->stmts.insert(op->As<ir::Block>()->stmts.begin() + i + 1, insert_node_);
```
Comment (zhhsplendid) [inline cinn/ir/ir_schedule.cc:1221]: Add brace for coding style.

Reference: https://google.github.io/styleguide/cppguide.html#Conditionals

File: cinn/ir/ir_schedule.cc [hunk 5]
```diff
@@ -1201,6 +1201,58 @@ Expr IRSchedule::CacheWrite(const Expr& block, int write_buffer_index, const std
...
+  }
+
+ private:
+  const Expr& ir_node_;
+  const Expr& insert_node_;
+  bool after_node_;
+};
+
+void IRSchedule::SyncThreads(const Expr& ir_node, bool after_node) {
+  CHECK(ir_node.As<ScheduleBlockRealize>() || ir_node.As<ir::For>());
```
Comment (zhhsplendid) [inline cinn/ir/ir_schedule.cc:1249]: Will we want `ir_node.As<ir::Block>()`?

File: cinn/backends/ir_schedule_test.cc [hunk 6]
```diff
@@ -1464,6 +1464,10 @@ TEST(IrSchedule, cache_read3) {
   auto a_cache = ir_sch.CacheRead(block_b, 0, "local");
   auto block_c = ir_sch.GetBlock("C");
   auto b_cache = ir_sch.CacheRead(block_c, 0, "local");
+  auto loops_c = ir_sch.GetLoops("C");
```
Comment (zhhsplendid) [inline cinn/backends/ir_schedule_test.cc:1467]: Should we test both `loop` and `block` ?

File: cinn/ir/ir_schedule.cc (hunk 1 above)
Comment (haozech) [reply cinn/ir/ir_schedule.cc]: Done, thanks

File: cinn/ir/ir_schedule.h (hunk 2 above)
Comment (haozech) [reply cinn/ir/ir_schedule.h]: Done, thanks

File: cinn/ir/ir_schedule.cc (hunk 3 above)
Comment (haozech) [reply cinn/ir/ir_schedule.cc:1204]: Ok, I plan to move all IR utility codes to a new file in my next PR.

File: cinn/ir/ir_schedule.cc (hunk 4 above)
Comment (haozech) [reply cinn/ir/ir_schedule.cc:1221]: Done, thanks

File: cinn/ir/ir_schedule.cc (hunk 5 above)
Comment (haozech) [reply cinn/ir/ir_schedule.cc:1249]: No, since there is no api for users to get a `ir::Block` node.

File: cinn/backends/ir_schedule_test.cc (hunk 6 above)
Comment (haozech) [reply cinn/backends/ir_schedule_test.cc:1467]: New test added. Thanks

**Review by SunNy820828449 — APPROVED** [review]
LGTM

**Pull request merged**